### PR TITLE
Add prisms auto-derivation

### PR DESCRIPTION
**Pull request by miguelangel-dev**
## Goal

Right now, in prisms, if the focus has associated values, we need to provide a bit of help by providing a pattern-matching function that extracts the associated values out of the case we are focusing on. 

Following the approach seen in [Point-Free](https://www.pointfree.co/episodes/ep89-case-paths-for-free) - I have updated it to auto derivate using reflection. So, no more boilerplate will be needed.

## Implementation details

Remove boilerplate in prisms using reflection for extracting the associated values from a `case`.

## Testing details

Added some tests for covering the next corner cases:
- Prism auto-derivation must support 2 cases: enums with/without associated types.
- Prism auto-derivation must support associated types with/without labelled.
- Prism auto-derivation must extract associated types in cases with the same name but with a different labelled parameter.

### CODE REVIEW

File: Tests/BowOpticsTests/PrismTest.swift [hunk 1]
```diff
@@ -122,4 +122,42 @@ class PrismTest: XCTestCase {
...
+    }
+    
+    enum EnumLabeled: AutoPrism {
+        case labeled(label: String)
+        case labeled(anotherLabel: String)
+    }
+    
+    func testAutoDerivatePrism_WithoutAssociatedTypes() {
+        let prism = Authentication.prism(for: Authentication.unkown)
+        XCTAssert(prism.getOption(.unkown).isDefined)
```
Comment (truizlop) [inline Tests/BowOpticsTests/PrismTest.swift]:
```suggestion
        XCTAssertTrue(prism.getOption(.unkown).isDefined)
```

File: Tests/BowOpticsTests/PrismTest.swift [hunk 2]
```diff
@@ -122,4 +122,42 @@ class PrismTest: XCTestCase {
...
+    }
+    
+    func testAutoDerivatePrism_WithAssociatedTypes() {
+        let prism = Authentication.prism(for: Authentication.authorized)
+        let result = prism.getOption(.authorized(8, "information")).toOptional() ?? (0, "")
+        XCTAssert(result == (8, "information"))
```
Comment (truizlop) [inline Tests/BowOpticsTests/PrismTest.swift]:
```suggestion
        XCTAssertTrue(result == (8, "information"))
```

File: Tests/BowOpticsTests/PrismTest.swift [hunk 3]
```diff
@@ -122,4 +122,42 @@ class PrismTest: XCTestCase {
...
+    }
+    
+    func testAutoDerivatePrism_WithLabeledAssociatedTypes() {
+        let prism = Authentication.prism(for: Authentication.requested)
+        let result = prism.getOption(.requested(1, info: "information")).toOptional() ?? (0, "")
+        XCTAssert(result == (1, "information"))
```
Comment (truizlop) [inline Tests/BowOpticsTests/PrismTest.swift]:
```suggestion
        XCTAssertTrue(result == (1, "information"))
```

File: Sources/BowOptics/Prism.swift [hunk 4]
```diff
@@ -421,19 +421,6 @@ public extension Prism where S == A, S == T, A == B {
...
-    /// - Parameters:
-    ///   - extract: Function to extract a value from the source.
-    ///   - embed: Function to embed a value into the source.
-    convenience init(extract: @escaping (S) -> A?, embed: @escaping (A) -> S) {
-        self.init(
-            getOrModify: { s in extract(s).flatMap(Either.right) ?? .left(s) },
-            reverseGet: embed)
-    }
-}
-
```
Comment (truizlop) [inline Sources/BowOptics/Prism.swift]: This constructor is still useful for cases where you want to build the Prism yourself. Please, revert this change.

Comment (miguelangel-dev) [reply Sources/BowOptics/Prism.swift]: Sure! Make sense :)
